### PR TITLE
Allow flaky conditions to inspect failure exceptions

### LINKED ISSUE #230: For "only_rerun" allow access exception attributes
I often have cases like below when I do not fail a test based on a specific exception attributes.

In case of `pytest-rerunfailures` the logic can look like:
```py
@pytest.mark.flaky(
    only_rerun=ClientResponseError,
    condition=lambda error: error.status in {429, 500},
)
async def test_request():
    await request("www.example.com")



@pytest.mark.flaky(
    only_rerun=CustomDatabaseError,
    condition=lambda error: error["code"] != 123456,
)
async def test_database():
    await database.set("foo", "bar")



def _callback(error: Exception) -> bool:
    if isinstance(error, CustomDatabaseError): ...
    if isinstance(error, AnotherError): ...
    raise 

@pytest.mark.flaky(
    only_rerun=[CustomDatabaseError, AnotherError],
    condition=_callback,
)
async def test_database():
    await database.set("foo", "bar")
```

There are also some 3rd-party packages that raise custom exceptions and I must handle an exception in different ways based on some attributes.

Question:  
Is this functionality out of scope of the current plugin and I should write a separate plugin?

### PR DESCRIPTION
**Pull request by ryux1**
Closes #230

This lets flaky conditions make one rerun decision for an entire test attempt from the exceptions that caused its failed phases:

- one-argument callable conditions receive each failure exception;
- existing zero-argument callable conditions remain supported;
- string conditions inspect the exception through the reserved `error` name;
- setup, call, teardown, and subtest failures retain their corresponding exceptions;
- an attempt is rerun when any failed phase matches, with each failure evaluated at most once.

`ExceptionInfo` stays on the worker-local test item, so xdist never serializes it, and the cached exception state is cleared after every attempt. Condition evaluation errors become non-fatal warnings even when warning filters promote warnings to errors.

Validation:

- pytest 8.2, 8.3, and 8.4: 206 passed, 19 skipped in each environment
- pytest 9.0 and 9.1: 216 passed, 9 skipped in each environment
- pytest 9.1 with pytest-xdist: 225 passed
- `pre-commit run --all-files` passed

### CODE REVIEW

**Comment by icemac** [comment]
The test failures against pytest-main are also on `master` and need to be fixed there before we can make progress here.

**Comment by icemac** [comment]
`master` is fixed now, could you please resolve the conflicts?

File: src/pytest_rerunfailures.py [hunk 1]
```diff
@@ -317,6 +337,7 @@ def evaluate_condition(item, mark, condition: object) -> bool:
             "sys": sys,
             "platform": platform,
             "config": item.config,
+            "error": error,
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **A string condition that raises now aborts the entire run.**

(The actual code is a bit further down, in the `except Exception as exc` block of the string branch — around line 362 — which is outside the diff, so I'm anchoring here.)

That handler still ends in `fail(..., pytrace=False)`. Since `_should_not_rerun` is now reached from `pytest_runtest_protocol`, i.e. outside any test phase, the `Failed` exception escapes the hook:

```python
@pytest.mark.flaky(reruns=1, condition="error.status == 429")
def test_x():
    assert False
```

```
INTERNALERROR> Failed: Error evaluating 'flaky' condition
INTERNALERROR>     error.status == 429
INTERNALERROR> AttributeError: 'AssertionError' object has no attribute 'status'
no tests ran
```

This is not an edge case: introducing `error` makes eval-time `AttributeError` the *normal* outcome, because the attribute being probed only exists on some exception types. The callable branch handles the identical input gracefully (warn, return `False`); the string branch should do the same.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 2]
```diff
@@ -294,21 +294,41 @@ def get_reruns_delay_backoff_factor(item):
...
+    # Callable condition.
+    if callable(condition):
+        try:
+            return bool(condition(error))
+        except Exception as exc:
+            msglines = [
+                f"Error evaluating {mark.name!r} condition as a callable",
+                *traceback.format_exception_only(type(exc), exc),
+            ]
+            warnings.warn("\n".join(msglines))
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **This `warnings.warn` also crashes the run under `filterwarnings = error`.**

`_should_not_rerun` runs inside `pytest_runtest_protocol`, so the `UserWarning` is raised (not recorded) and propagates out of the hook. Your own `test_callable_condition_error_prevents_rerun` scenario plus a `pytest.ini` with `filterwarnings = error`:

```
INTERNALERROR> UserWarning: Error evaluating 'flaky' condition as a callable
INTERNALERROR> ValueError: nope
```

`filterwarnings = error` is a very common setting. Consider reporting via `item.warn(...)` or catching the raised warning here.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 3]
```diff
@@ -1043,6 +1077,13 @@ def pytest_runtest_makereport(item, call):
         # create a dict to store xfail results for each stage
         setattr(item, "_test_xfailed", {})
 
+        # Keep exception state on the worker-side item. TestReport attributes
+        # are serialized by pytest-xdist and ExceptionInfo is not serializable.
+        setattr(item, "_rerun_condition_excinfo", {})
+        setattr(item, "_rerun_condition_results", {})
+
+    item._rerun_condition_excinfo[result.when] = call.excinfo
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **This unconditional assignment clobbers subtest exceptions, so the feature never fires for failed subtests.**

With `pytest-subtests`, `pytest_runtest_makereport` fires twice for `when == "call"`: first for the subtest report (real `excinfo`), then for the aggregate report where `call.excinfo is None`. Verified with a trylast wrapper:

```
MAKEREPORT when=call failed=True  excinfo=<ExceptionInfo ValueError('boom') tblen=1>
MAKEREPORT when=call failed=False excinfo=None
```

So `condition=lambda error: isinstance(error, ServiceError) and error.status == 429` on a test whose subtest raises `ServiceError(429)` sees `error=None` and is **not** re-run — even though re-running on failed subtests is explicitly supported (`has_failed_subtests`, #357).

The `_test_xfailed` line right below already solves this by accumulating rather than replacing; the same shape works here:

```python
if call.excinfo is not None or result.when not in item._rerun_condition_excinfo:
    item._rerun_condition_excinfo[result.when] = call.excinfo
```

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 4]
```diff
@@ -565,18 +586,20 @@ def _should_hard_fail_on_error(item, report, excinfo):
...
         or (not report.failed and not has_failed_subtests)
         or xfail
         or is_terminal_error
-        or not condition
-    )
+    ):
+        return True
+
+    excinfo = item._rerun_condition_excinfo.get(report.when)
+    return not get_reruns_condition(item, excinfo, report.when)
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **Evaluating the condition per phase lets the phases disagree, and a test can end up reported as both FAILED and PASSED.**

The condition result is now keyed by `report.when`, so the call phase can answer "don't re-run" while the teardown phase answers "re-run". The call report is then logged as a final failure, and the teardown report still triggers the re-run. With a call phase failing on `AssertionError`, a function-scoped fixture whose teardown raises `ServiceError(503)`, and `condition=lambda e: getattr(e, "status", None) == 503`:

```
test_a.py::test_x FAILED
test_a.py::test_x RERUN
test_a.py::test_x PASSED
1 failed, 2 passed, 1 rerun
```

Before this PR the condition was a single item-wide value, so the two reports could not disagree. Related: because `pytest_runtest_teardown` evaluated the *call*-phase condition as falsy, the higher-scope finalizers were not held back for a re-run that then happened anyway — the exact situation #357 was fixed to avoid.

Worth deciding explicitly what the intended semantics are: one decision per attempt (e.g. "re-run if any failed phase's condition matches"), rather than an independent decision per report.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 5]
```diff
@@ -317,6 +337,7 @@ def evaluate_condition(item, mark, condition: object) -> bool:
             "sys": sys,
             "platform": platform,
             "config": item.config,
+            "error": error,
         }
         if hasattr(item, "obj"):
             globals_.update(item.obj.__globals__)  # type: ignore[attr-defined]
```
Comment (icemac) [inline src/pytest_rerunfailures.py:385]: **This `update()` silently shadows the new `error` name.**

`globals_["error"]` is set three lines up and then overwritten by anything named `error` in the test module (a variable, or `from x import error`):

```python
error = None  # module level

@pytest.mark.flaky(reruns=1, condition="isinstance(error, ServiceError)")
def test_x():
    raise ServiceError(429)
```

→ `1 failed`, no re-run; the condition saw the module global. Setting `"error"` *after* the `update()` makes the documented name unshadowable.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 6]
```diff
@@ -294,21 +294,41 @@ def get_reruns_delay_backoff_factor(item):
...
 
 
-def evaluate_condition(item, mark, condition: object) -> bool:
+def evaluate_condition(item, mark, condition: object, excinfo=None) -> bool:
     # copy from python3.8 _pytest.skipping.py
 
+    error = excinfo.value if excinfo is not None else None
+
+    # Callable condition.
+    if callable(condition):
```
Comment (icemac) [inline src/pytest_rerunfailures.py:346]: **This is a breaking change for existing zero-argument callable conditions, and it fails silently.**

Checking `callable(condition)` before the boolean branch changes behaviour for anyone who passed a function today: it used to fall through to `bool(condition)` — always truthy, i.e. "always re-run". Now it is invoked with one positional argument.

```python
@pytest.mark.flaky(reruns=1, condition=lambda: sys.platform == "darwin")
```

- master: `1 passed, 1 rerun`
- this branch: `1 failed, 1 warning` — the `TypeError` is swallowed into a generic warning and re-runs are silently off.

Two suggestions:
1. The changelog fragment is `changes/230.feature.rst`; towncrier is configured with a `breaking` category, which this warrants.
2. An arity mismatch deserves a specific message ("condition callable must accept one argument") rather than the same generic warning used for exceptions raised *inside* the condition — otherwise the failure mode is hard to diagnose.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 7]
```diff
@@ -927,6 +950,16 @@ def _is_rerun_path_excluded(item):
     )
 
 
+def _get_reruns_condition_failure(item):
+    """Return the phase and exception for the most recent failed test phase."""
+    failed_statuses = getattr(item, "_test_failed_statuses", {})
+    excinfos = getattr(item, "_rerun_condition_excinfo", {})
+    for phase in ("teardown", "call", "setup"):
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **The `"teardown"` iteration is unreachable, and the `(None, None)` return defeats the caching this PR adds.**

This helper is only called from `pytest_runtest_teardown`, which runs *before* the teardown report is created, so `_test_failed_statuses["teardown"]` is never set at this point.

When only subtests failed (the call phase itself passing), it returns `(None, None)`, and `get_reruns_condition` with `phase=None` doesn't consult or populate the cache — so the condition runs twice in a single attempt. With a counting callable condition on a failing subtest, `COND ... None` is printed twice per attempt, which contradicts the "each phase condition is evaluated once" claim in the PR description.

Minor: the docstring says "most recent failed test phase", but the order here is fixed, not chronological.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py [hunk 8]
```diff
@@ -1043,6 +1077,13 @@ def pytest_runtest_makereport(item, call):
         # create a dict to store xfail results for each stage
         setattr(item, "_test_xfailed", {})
 
+        # Keep exception state on the worker-side item. TestReport attributes
+        # are serialized by pytest-xdist and ExceptionInfo is not serializable.
+        setattr(item, "_rerun_condition_excinfo", {})
```
Comment (icemac) [inline src/pytest_rerunfailures.py]: **These dicts keep an `ExceptionInfo` alive on every item for the whole session.**

`_rerun_condition_excinfo` is populated for *all* items, flaky or not, and is only reset at the next `setup` of the same item — which never happens for the final attempt. Since `session.items` holds every item until the run ends, each failed test's full traceback (and therefore all of its frame locals) stays reachable. pytest deliberately keeps only the string `longrepr` on reports for exactly this reason.

Clearing the dict once the item's re-run decision is final (or only recording it when a flaky marker with a `condition` is present) would avoid the retention.

— _Comment created by Claude_

File: docs/mark.rst [hunk 9]
```diff
@@ -56,6 +58,36 @@ The condition can be any expression that evaluates to ``True`` or ``False``.
...
+Its evaluation context also contains ``os``, ``sys``, ``platform``, ``config``
+(the pytest config object), and the test function's globals:
+
+.. code-block:: python
+
+   @pytest.mark.flaky(reruns=3, condition="error.status in {429, 503}")
+   def test_service_request():
+       raise TemporaryError(429)
+
+If a callable condition raises an exception, pytest emits a warning and does
```
Comment (icemac) [inline docs/mark.rst]: This sentence doesn't hold in two cases, both of which produce `INTERNALERROR` rather than a warning: string conditions that raise during evaluation, and callable conditions under `filterwarnings = error` (see the two inline comments in `src/pytest_rerunfailures.py`). Once those are fixed the docs will be accurate as written.

Separately, the `error` name is documented for string conditions without mentioning that a module-level `error` in the test file overrides it.

— _Comment created by Claude_

**Review by icemac — COMMENTED** [review]
Thanks for working on #230 — the xdist-safe approach of keeping `ExceptionInfo` on the worker-side item is the right idea. A few things need fixing before this can go in; two of them abort the whole run with `INTERNALERROR`, and one makes the feature a no-op for failed subtests. Details inline.

— _Comment created by Claude_

File: src/pytest_rerunfailures.py (hunk 1 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef makes string evaluation failures non-fatal warnings and returns false for that failure. The regression covers a missing exception attribute and asserts that pytest does not exit with INTERNAL_ERROR.

File: src/pytest_rerunfailures.py (hunk 2 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef routes condition warnings through a helper that catches warning exceptions, so `filterwarnings = error` cannot escape the protocol hook. The new regression exercises that exact configuration.

File: src/pytest_rerunfailures.py (hunk 3 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef accumulates only failing, non-null ExceptionInfo values per phase instead of overwriting them. The new pytest-subtests regression verifies the callable receives ServiceError(429) exactly once.

File: src/pytest_rerunfailures.py (hunk 4 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef now computes one attempt-level decision after all reports: any matching failed phase reruns the attempt. Nonmatching earlier failures are not published as final when a later phase carries the rerun, and finalizers remain suspended until the complete decision is known. The mixed call/teardown regression exits cleanly with one pass and one rerun.

File: src/pytest_rerunfailures.py (hunk 5 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py:385]: 04cb3ef applies test globals first and then assigns the reserved `error` binding. The regression defines a module-level `error = None` and verifies the actual ServiceError still drives the rerun.

File: src/pytest_rerunfailures.py (hunk 7 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef replaces the phase-selection helper with an accumulated failed-exception list and one cached attempt decision. The subtest regression verifies its exception is evaluated once, and teardown failures are included once their report exists.

File: src/pytest_rerunfailures.py (hunk 8 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py]: 04cb3ef clears both the ExceptionInfo collection and its result cache at the end of every attempt. A regression inspects the collected item from the following test and verifies both stores are empty.

File: docs/mark.rst (hunk 9 above)
Comment (ryux1) [reply docs/mark.rst]: 04cb3ef updates the docs to cover callable and string failures, the reserved unshadowable `error` binding, zero-argument compatibility, and the any-matching-failure attempt semantics. Both INTERNAL_ERROR cases now have regressions.

File: src/pytest_rerunfailures.py (hunk 6 above)
Comment (ryux1) [reply src/pytest_rerunfailures.py:346]: 04cb3ef preserves zero-argument callable conditions by inspecting the callable signature; one-argument callables receive the exception. Unsupported arities get a specific zero-or-one-argument warning. The changelog remains a feature because the existing zero-argument behavior is retained.

**Review by icemac — APPROVED** [review]
